### PR TITLE
Added platform APIs for handling resets.

### PR DESCRIPTION
**Pull request by darconeous**
This change includes the addition of two new platform API functions:

* `otPlatReset()`
* `otPlatGetResetReason()`

They are pretty straightforward. Every platform must implement both, but
`otPlatReset()` is allowed to simply return if the platform doesn't support
being reset (like the posix platform).

This commit addresses issue #283.

### CODE REVIEW

**Comment by mszczodrak** [comment]
LGTM

**Comment by darconeous** [comment]
I want to make a few more changes before we merge.

File: src/ncp/ncp_base.cpp [hunk 1]
```diff
@@ -868,9 +915,15 @@ void NcpBase::CommandHandler_NOOP(uint8_t header, unsigned int command, const ui
 
 void NcpBase::CommandHandler_RESET(uint8_t header, unsigned int command, const uint8_t *arg_ptr, uint16_t arg_len)
 {
-    // TODO: Figure out how to actually perform a reset.
-    otInit();
-    SendLastStatus(SPINEL_HEADER_FLAG | SPINEL_HEADER_IID_0, SPINEL_STATUS_RESET_SOFTWARE);
+    otPlatReset();
+
+    // We only get to this point if the
```
Comment (abtink) [inline src/ncp/ncp_base.cpp]: Here we set all bits of `mChangedFlags`? For example, we are setting both `OT_IP6_ADDRESS_ADDED` and `OT_IP6_ADDRESS_REMOVED` flags. This works ok now (we clear them all together as we send stuff out).  I feel we may need to be careful about the value we set here when we add new flags (e.g., if we handle INSERTED case).

Comment (darconeous) [inline src/ncp/ncp_base.cpp]: I just changed this behavior in a recent commit.

**Comment by darconeous** [comment]
I think this is good to go.

**Comment by darconeous** [comment]
Oops, there was a pretty check failure. Will fix.

**Comment by darconeous** [comment]
Ok, fixed.

**Comment by abtink** [comment]
LGTM :+1:

**Comment by jwhui** [comment]
LGTM 👍

**Comment by jwhui** [comment]
@darconeous, #278 caused a merge conflict.  Sorry about that.

**Comment by darconeous** [comment]
Whomp-Whomp. Will fix.

**Comment by darconeous** [comment]
Should be fixed now.

**Comment by darconeous** [comment]
@jwhui Feel free to merge when you can.

**Pull request merged**